File: mobile/ios/Runner/Services/staking/StakingService.swift
//
//  StakingService.swift
//  TigerWallet
//
//  Complete Staking Service for iOS
//

import Foundation

class StakingService {
    static let shared = StakingService()
    private let baseURL = "https://api.tigerwallet.com/v1/staking"
    
    private init() {}
    
    func getChains(completion: @escaping (Result<[StakingChain], Error>) -> Void) {
        let url = URL(string: "\(baseURL)/chains")!
        
        URLSession.shared.dataTask(with: url) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data else {
                completion(.failure(NSError(domain: "StakingService", code: -1, userInfo: [NSLocalizedDescriptionKey: "No data"])))
                return
            }
            
            do {
                let chains = try JSONDecoder().decode([StakingChain].self, from: data)
                completion(.success(chains))
            } catch {
                completion(.failure(error))
            }
        }.resume()
    }
    
    func getValidators(chain: String, completion: @escaping (Result<[Validator], Error>) -> Void) {
        let url = URL(string: "\(baseURL)/\(chain)/validators")!
        
        URLSession.shared.dataTask(with: url) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data else {
                completion(.failure(NSError(domain: "StakingService", code: -1, userInfo: [NSLocalizedDescriptionKey: "No data"])))
                return
            }
            
            do {
                let validators = try JSONDecoder().decode([Validator].self, from: data)
                completion(.success(validators))
            } catch {
                completion(.failure(error))
            }
        }.resume()
    }
    
    func stake(chain: String, validatorAddress: String, amount: String, address: String, completion: @escaping (Result<String, Error>) -> Void) {
        let url = URL(string: "\(baseURL)/\(chain)/stake")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = [
            "validatorAddress": validatorAddress,
            "amount": amount,
            "address": address
        ]
        
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let txHash = json["txHash"] as? String else {
                completion(.failure(NSError(domain: "StakingService", code: -1, userInfo: [NSLocalizedDescriptionKey: "Stake failed"])))
                return
            }
            
            completion(.success(txHash))
        }.resume()
    }
    
    func unstake(chain: String, validatorAddress: String, amount: String, address: String, completion: @escaping (Result<String, Error>) -> Void) {
        let url = URL(string: "\(baseURL)/\(chain)/unstake")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = [
            "validatorAddress": validatorAddress,
            "amount": amount,
            "address": address
        ]
        
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let txHash = json["txHash"] as? String else {
                completion(.failure(NSError(domain: "StakingService", code: -1, userInfo: [NSLocalizedDescriptionKey: "Unstake failed"])))
                return
            }
            
            completion(.success(txHash))
        }.resume()
    }
    
    func claimRewards(chain: String, validatorAddress: String, address: String, completion: @escaping (Result<String, Error>) -> Void) {
        let url = URL(string: "\(baseURL)/\(chain)/claim")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = [
            "validatorAddress": validatorAddress,
            "address": address
        ]
        
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let txHash = json["txHash"] as? String else {
                completion(.failure(NSError(domain: "StakingService", code: -1, userInfo: [NSLocalizedDescriptionKey: "Claim failed"])))
                return
            }
            
            completion(.success(txHash))
        }.resume()
    }
    
    func getPositions(address: String, chain: String, completion: @escaping (Result<[StakingPosition], Error>) -> Void) {
        let url = URL(string: "\(baseURL)/\(chain)/positions/\(address)")!
        
        URLSession.shared.dataTask(with: url) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data else {
                completion(.failure(NSError(domain: "StakingService", code: -1, userInfo: [NSLocalizedDescriptionKey: "No data"])))
                return
            }
            
            do {
                let positions = try JSONDecoder().decode([StakingPosition].self, from: data)
                completion(.success(positions))
            } catch {
                completion(.failure(error))
            }
        }.resume()
    }
}

struct StakingChain: Codable {
    let id: String
    let name: String
    let symbol: String
    let stakingAPY: Double
    let minStake: String
    let lockPeriod: Int
}

struct Validator: Codable {
    let address: String
    let name: String
    let commission: Double
    let apy: Double
    let totalStaked: String
    let delegators: Int
    let uptime: Double
}

struct StakingPosition: Codable {
    let validatorAddress: String
    let stakedAmount: String
    let rewardsAccrued: String
    let rewardsClaimed: String
    let chain: String
    let status: String
}

File: browser_extensions/chrome/src/services/staking-service.js
/**
 * TigerWallet Browser Extension - Staking Service
 */

class StakingService {
    constructor() {
        this.apiBase = 'https://api.tigerwallet.com/v1/staking';
    }

    // Get staking chains
    async getChains() {
        const response = await fetch(`${this.apiBase}/chains`);
        if (!response.ok) throw new Error('Failed to fetch chains');
        return response.json();
    }

    // Get validators for a chain
    async getValidators(chain) {
        const response = await fetch(`${this.apiBase}/${chain}/validators`);
        if (!response.ok) throw new Error('Failed to fetch validators');
        return response.json();
    }

    // Get validator details
    async getValidator(chain, validatorAddress) {
        const response = await fetch(`${this.apiBase}/${chain}/validator/${validatorAddress}`);
        if (!response.ok) throw new Error('Failed to fetch validator');
        return response.json();
    }

    // Stake
    async stake(chain, validatorAddress, amount, address) {
        const response = await fetch(`${this.apiBase}/${chain}/stake`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ validatorAddress, amount, address })
        });
        if (!response.ok) throw new Error('Failed to stake');
        return response.json();
    }

    // Unstake
    async unstake(chain, validatorAddress, amount, address) {
        const response = await fetch(`${this.apiBase}/${chain}/unstake`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ validatorAddress, amount, address })
        });
        if (!response.ok) throw new Error('Failed to unstake');
        return response.json();
    }

    // Claim rewards
    async claimRewards(chain, validatorAddress, address) {
        const response = await fetch(`${this.apiBase}/${chain}/claim`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ validatorAddress, address })
        });
        if (!response.ok) throw new Error('Failed to claim rewards');
        return response.json();
    }

    // Get staking positions
    async getPositions(address, chain) {
        const response = await fetch(`${this.apiBase}/${chain}/positions/${address}`);
        if (!response.ok) throw new Error('Failed to fetch positions');
        return response.json();
    }

    // Get pending unstaking
    async getPendingUnstakes(address, chain) {
        const response = await fetch(`${this.apiBase}/${chain}/pending/${address}`);
        if (!response.ok) throw new Error('Failed to fetch pending');
        return response.json();
    }

    // Get rewards history
    async getRewardsHistory(address, chain) {
        const response = await fetch(`${this.apiBase}/${chain}/rewards/${address}`);
        if (!response.ok) throw new Error('Failed to fetch rewards');
        return response.json();
    }

    // Calculate APY
    calculateAPY(rewardRate, compoundingPeriods = 365) {
        return Math.pow(1 + rewardRate, compoundingPeriods) - 1;
    }
}

window.TigerWalletStakingService = new StakingService();
